만국 박람회 [STEP 3] Lingo, 마이노

seu0313 opened the pull request:
@wonhee009 
안녕하세요 라자냐 👍
만국 박람회 [STEP 3] PR 보냅니다. 잘 부탁드립니다!

### 고민 및 해결한 점

**1. Alert**
범용성만을 위해 모든 기능을 구현하는 것은 오히려 개발자(팀원들)에게 혼란을 줄 수 있기 때문에 필요한 기능만 구현하도록 해야한다는 점에 대해 고민했습니다.

**2. 첫 화면은 portrait으로 고정, 다른 화면은 landScape 허용**
AppDelegate의 해당 메서드를 통해 MainViewController의 viewWillAppear에서 UIInterfaceOrientationMask을 portrait으로 고정 시켜준 뒤 viewWillDisappear에서 다시 모든 화면 전환이 가능하도록 고정을 풀어주어 해결하였습니다.

```swift
func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?)
```

**3. Dynamic Type를 적용하여 사용자의 접근성을 높였습니다.**
Dynamic Type을 적용하는 과정에서 Label의 Font Size가 너무 커져버려서 글자가 짤리는 현상이 발생하는 Label들에 대해 adjustsFontSizeToFitWidth 프로퍼티의 값을 true로 설정하여 해결하였습니다.

**4. AttributedString**
`AttributedString`을 사용하여 텍스트의 특정 범위에 폰트를 다르게 줄 수 있다는 것을 알게되었고 공식문서를 참고하여 적용해봤습니다.

**5. ImageView의 사이즈에 대하여**
ImageView의 width를 설정해주면 자동으로 height를 원본 이미지 비율에 맞게 설정해준다고 생각했었습니다. 오토레이아웃을 잡아주며 공부해보니 ImageView는 처음에는 Intrinsic Size가 없다가 이미지를 넣어주면 이미지의 크기를 Intrinsic Size로 설정하는 것이었고 크기는 AspectFit, AspectFill과 같은 Content Mode에 따라 변경될 수 있다는 것을 알게되었습니다.

---

### 조언 받고 싶은 점
**1. Button titleLabel Dynamic Type 적용**
Button의 titleLabel의 Dynamic Type을 적용했을 때 FontSize를 컨트롤하는 부분을 titleLabel의 adjust관련 프로퍼티들과 autoShrink관련 프로퍼티들로 시도해보았으나 아직 해결하지 못하였습니다. 버튼의 경우 어떤식으로 해결하는것이 좋을까요?

**2. LandScape 상태에서 MainVC -> ExpoItemTableVC로 push되는 상황의 화면 전환**
현재 MainVC에서 landScape 모드로 다음 VC로 push 되는 경우에 자동으로 화면이 전환되지 않고 portrait 모드로 push가 된 뒤, 시뮬레이터의 디바이스 방향을 전환해야 landScape 모드로 전환되는 상황입니다. ExpoItemTableVC에서 MainVC처럼 강제로 방향을 전환하면 해결은 할 수 있지만, landScape의 방향(left, right)를 모르기 때문에 만약 디바이스의 방향이 landScapeLeft인 경우에 landScapeRight로 강제로 전환할 경우 같은 문제가 있습니다.

현재 디바이스의 방향을 확인해서 해당 방향으로 강제로 전환 시켜주는 방법이 있을까요?

Commit:
refactor: AlertBuilder 제거 및 각 VC에 Alert 생성 기능 추가

Mino777 committed:
feat: MainViewController 화면전환 Portrait으로 고정 구현

Commit:
feat: MainViewController에 Dynamic Text 및 AttributedString 적용

Mino777 committed:
feat: MainViewController AutoLayout 추가

Commit:
refactor: ExpoItems 테이블 뷰에 Dynamic Type 적용 및 오토레이아웃 수정

Mino777 committed:
refactor: ExpoItemDetailViewController Auto Layout 수정

Commit:
docs: STEP3까지의 PR 내용을 추가

wonhee009 commented on Expo1900/Expo1900/Controller/ExpoItemTableViewController.swift:
```diff
@@ -33,12 +32,32 @@ private extension ExpoItemTableViewController {
+      alertController.addAction(action)
     }
+    self.present(alertController, animated: true)
   }
```
`showAlertController`라는 메서드가 중복 구현되어 있는거 같은데요.
각자의 VC가 Alert을 생성하는 책임을 갖는것은 좋지만 중복은 피해야하지 않을까요?👀
`AlertController`만 만들어주는 class/struct를 구현하고 VC에서는 해당 class/struct를 통해 alert를 생성하는 것도 방법일거 같습니다.

seu0313 replied:
AlertControllerable 프로토콜에서 extension에 기본 구현하여 중복되지 않도록 분리했습니다. AlertControllerable 프로토콜을 채택한 ViewController만 얼럿을 사용할 수 있도록 했습니다.

wonhee009 commented on Expo1900/Expo1900/Controller/MainViewController.swift:
```diff
@@ -52,22 +57,85 @@ private extension MainViewController {
+    let attributedString = NSMutableAttributedString(string: text)
+    attributedString.addAttribute(.font, value: font, range: range)
+    label.attributedText = attributedString
+  }
```
> AttributedString을 사용하여 텍스트의 특정 범위에 폰트를 다르게 줄 수 있다는 것을 알게되었고 공식문서를 참고하여 적용

폰트를 다르게 주는게 명세였을까요?

![image](https://user-images.githubusercontent.com/30456206/164226074-e5aeea89-40ba-4c4a-9db6-123e0cbb5851.png)

seu0313 replied:
프로젝트 STEP 3 요구 조건에서 Dynamic Type이 XXXL일 경우 어떻게 되어야한다는 명세는 없었지만

기존에는 XXXL에서 아래의 사진처럼 크기가 다른 문제가 있었습니다. 따라서, AttributedString을 사용하던 방식을 두개의 Label로 분리하여 Dymamic Type xxxLarge에서 발생하던 문제를 해결했습니다. 

|**기존**|**수정**|
|:---:|:---:|
|<img width="200" src="https://user-images.githubusercontent.com/94151993/164362862-7712d644-c544-468a-838a-3ab101a71ebb.png"/>|<img width="200" src="https://user-images.githubusercontent.com/94151993/164363060-a8871313-0735-47c9-a73f-a3fbf1c6078d.png"/>|

wonhee009 replied:
iPhone SE 기준 확인 부탁드려요!
![image](https://user-images.githubusercontent.com/30456206/164442318-308085f8-a1ff-4e0d-ab76-7148e5bd1dc9.png)

seu0313 replied:
`개최기간`의 Label의 Compression Resistance 를 변경하여 문제를 해결할 수 있엇습니다!

<img width="250px" src="https://user-images.githubusercontent.com/94151993/164584915-eb8ec35a-7021-4a8a-99ff-4699f9b68125.png"/>

wonhee009 commented on Expo1900/Expo1900/Controller/MainViewController.swift:
```diff
@@ -52,22 +57,85 @@ private extension MainViewController {
+  ) {
+    self.lockOrientation(orientation)
+    UIDevice.current.setValue(rotateOrientation.rawValue, forKey: Constants.orientation)
   }
```
디바이스 방향에 따른 처리를 잘해주었네요 👍 
이후 스펙에서 ExpoItemDetail 화면에서 사진 tap했을때 새로운 화면 A가 추가되었습니다.
화면 A는 세로로만 노출되어야 한다면 A ViewController에도 해당 메서드가 추가되어야 하는걸까요?
이 부분 생각해보면 좋을거 같습니다 😄

Mino777 replied:
AlertControllerable 처럼 protocol에 기본 구현을 통해 추후에 필요한 경우 LockOrientation protocol을 채택하여 사용할 수 있도록 분리해보았습니다!
감사합니다!

wonhee009 replied:
지금 방식도 좋습니다👍
`Lockorientation`이라는 네이밍이 어떤 역할의 프로토콜인지 와닿지 않는거 같습니다.

현재 MainVC > `viewWillAppear`, `viewWillDisappear`에서 lockOrientation을 호출해주고 있는데요.
portrait으로 lock하는 부분은 이해가 가지만 `viewWillDisappear`에서 .all로 lock하는 부분은 메서드의 네이밍과 안 맞지 않나 생각이 듭니다.

seu0313 replied:
.all 할때의 네이밍을 `setLockOrientation`으로 변경하여 설정한다는 의미로 변경했고 
LockOrientation 명사형을 Orientating 으로 변경하여 프로토콜의 네이밍에 더 맞도록 변경했습니다.

wonhee009 reviewed (changes requested):
@llingo @Mino777 스텝3 수고하셨습니다!
코드에 몇몇 코멘트 남겨두었으니 확인 부탁드려요 🙏 

1. 어떤 부분이 원하는 대로 안 되는지 좀 더 자세하게 알 수 있을까요?
2. "landScape의 방향(left, right)를 모르기 때문에" landScape로 강제 전환은 할 수 있지만 그 방향이 Left인지 Right인지를 모른다가 문제일까요?

Mino777 committed:
refactor: 기존 Alert 로직을 AlertControllerable protocol 기본 구현으로 분리

Commit:
refactor: 기존 디바이스의 Orientation을 설정을 LockOperation 프로토콜의 기본 구현으로 분리

Mino777 committed:
refactor: MainViewController Dynamic Type을 위한 Label 수정

Mino777 commented:
@wonhee009 
앗 저희가 좀 더 명확하게 질문을 드렸어야했는데 다시보니 너무 장황하네요

**1. Button의 경우 Content Size를 벗어나는 부분**
<img width="200" src="https://user-images.githubusercontent.com/54234176/164363014-d4e3e619-64e4-4c0b-8474-22412557fae3.png"/>
사진처럼 크기를 키웠을 때 버튼의 titleLabel이 버튼의 레이아웃을 벗어나는 부분의 해결법에 대한 질문이었습니다!

**2. "landScape의 방향(left, right)를 모르기 때문에" landScape로 강제 전환은 할 수 있지만 그 방향이 Left인지 Right인지를 모른다가 문제일까요?**

![Simulator Screen Shot - iPhone 12 mini - 2022-04-21 at 12 10 58](https://user-images.githubusercontent.com/54234176/164364455-ef843e47-7a2b-4f8f-881e-a8f7a1a6be5f.png)

위 사진 처럼 화면 방향을 전환한 상태로 진입했을 때 화면전환이 되지 않고, 시뮬레이터의 화면 방향을 전환 시켜야 스크린의 방향도 전환되는 상황입니다!

![Simulator Screen Shot - iPhone 12 mini - 2022-04-21 at 12 12 21](https://user-images.githubusercontent.com/54234176/164364400-03dde2c6-20cb-45fc-ad1a-4eec2bab920c.png)

그래서 만약에 두번째 화면에 진입하였을 때 강제로 landScapeRight로 전환을 시켜줄 수 있지만, 위 사진 처럼 방향이 반대 방향인 케이스에는 사진 같은 결과가 나오게 되는 부분에 대한 질문 이었습니다! 저희가 학습해본 결과 coreMotion을 통한 기울기 계산을 이용해 해결할 수 있는 부분인 것 같지만 이번 프로젝트에서 coreMotion 까지 사용하는게 맞는 걸까? 라는 생각에 우선 라자냐에게 여쭤보게 되었습니다!

seu0313 commented:
@wonhee009 

위 질문의 상황을 gif로 녹화한 것은 아래와 같습니다. 
> 업로드 최대 용량 문제로 화질이 깨져보이네요 😅 
이미지 클릭하시면 재생이 되는 것 같습니다 🙇🏻

![1111](https://user-images.githubusercontent.com/94151993/164367997-4f4c6f4a-b01e-4250-b0ff-1e4983c712c9.gif)

wonhee009 commented on Expo1900/Expo1900/Util/Protocol/LockOrientation.swift:
```diff
@@ -0,0 +1,32 @@
+  
+  func lockOrientation(
+    _ orientation: UIInterfaceOrientationMask,
+    andRotateTo rotateOrientation:UIInterfaceOrientation
```
```suggestion
    andRotateTo rotateOrientation: UIInterfaceOrientation
```

wonhee009 commented on Expo1900/Expo1900/Util/Protocol/AlertControllerable.swift:
```diff
@@ -0,0 +1,29 @@
+
+import UIKit
+
+protocol AlertControllerable {}
```
해당 프로토콜은 빈 프로토콜로 냅두놔요?
`extension`으로 기본 기능을 구현했지만 
```swift
protocol AlertControllerable { func showAlertController() }
```
로써 껍데기는 가지고 있어야하지 않을까요?

또한 `AlertControllerable`이라는 네이밍이 "AlertController가 될 수 있는"으로 느껴지는데요.
해당 네이밍을 만족하려면 프로토콜을 채택하는 곳이 AlertController의 기능을 할 수 있어야 할것 같습니다.
하지만 프로토콜에서 현재 해주고 있는 역할을 AlertController와는 다르지 않나 싶네요ㅎㅎ

seu0313 replied:
**얼럿을 보여줄 수 있는** 이라는 뜻으로 `AlertPresentable`로 변경했습니다.

wonhee009 commented on Expo1900/Expo1900/Util/Protocol/AlertControllerable.swift:
```diff
@@ -0,0 +1,29 @@
+protocol AlertControllerable {}
+
+extension AlertControllerable {
+  func showAlertController(
+    title: String?,
+    message: String?,
+    preferredStyle: UIAlertController.Style,
+    actions: [UIAlertAction]
+  ) -> UIAlertController {
```
`showAlertController`라고 하기에는 alert을 show하는 역할이 해당 메서드에는 없는거 같습니다.
좀 더 어울리는 네이밍 고려 부탁드려요!

seu0313 replied:
Alert 기능을 분리하는 과정에서 self.present 를 안하게 되었고 네이밍을 수정하지 못했던 것 같습니다.
`makeAlertController` 로 변경했습니다!

wonhee009 commented:
1. 버튼에 관한 부분은 Dynamic Type은 잘 적용되고 있으나 UI의 문제인거 같습니다. 버튼의 text의 높이가 정해진 이후 버튼의 높이를 다시 정의해줘야할 것 같은데요. 이 부분에 초점을 맞춰서 찾아보면 될 것 같습니다.
2. 음 현재는 coreMotion까지는 신경 쓰지 않아도 될거 같습니다 😄

Commit:
refactor: Alert, Orientation 프로토콜 네이밍 변경

AlertControllerable -> AlertPresentable, LockOrientation -> Orientating로 네이밍 변경

Mino777 committed:
refactor: Autolayout 수정

Button Dynamic Type, Label SE 대응 문제

seu0313 commented:
@wonhee009 

1. 버튼에 관한 부분은 Dynamic Type은 잘 적용되고 있으나 UI의 문제인거 같습니다. 버튼의 text의 높이가 정해진 이후 버튼의 높이를 다시 정의해줘야할 것 같은데요. 이 부분에 초점을 맞춰서 찾아보면 될 것 같습니다.

> 버튼이 포함된 StackView의 Alignment 속성을 Center로 변경해서 해결할 수 있었습니다. 

2. 음 현재는 coreMotion까지는 신경 쓰지 않아도 될거 같습니다 😄

> 네 감사합니다 👍

wonhee009 reviewed (approved):
스텝3 수고하셨습니다 😄
